File: csharp/GooseFeatures/MCPRegistry.cs
// Re-implementation of Goose's Model Context Protocol (MCP) Registry

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace TormentNexus.GooseFeatures
{
    public class MCPTool
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public string InputSchema { get; set; }
        public Func<string, Task<string>> Execute { get; set; }
    }

    public class MCPRegistry
    {
        private readonly Dictionary<string, MCPTool> tools = new Dictionary<string, MCPTool>();

        public void RegisterTool(MCPTool tool)
        {
            tools[tool.Name] = tool;
        }

        public MCPTool GetTool(string name)
        {
            tools.TryGetValue(name, out var tool);
            return tool;
        }

        public async Task<string> ExecuteTool(string name, string input)
        {
            if (tools.TryGetValue(name, out var tool))
            {
                return await tool.Execute(input);
            }
            throw new Exception($"Tool {name} not found in MCP registry.");
        }
    }
}

File: java/src/main/java/com/tormentnexus/goose/MCPRegistry.java
package com.tormentnexus.goose;

// Re-implementation of Goose's Model Context Protocol (MCP) Registry

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

public class MCPRegistry {

    public static class MCPTool {
        public String name;
        public String description;
        public String inputSchema;
        public Function<String, CompletableFuture<String>> execute;

        public MCPTool(String name, String description, String inputSchema, Function<String, CompletableFuture<String>> execute) {
            this.name = name;
            this.description = description;
            this.inputSchema = inputSchema;
            this.execute = execute;
        }
    }

    private final Map<String, MCPTool> tools = new HashMap<>();

    public void registerTool(MCPTool tool) {
        tools.put(tool.name, tool);
    }

    public MCPTool getTool(String name) {
        return tools.get(name);
    }

    public CompletableFuture<String> executeTool(String name, String input) {
        MCPTool tool = tools.get(name);
        if (tool != null) {
            return tool.execute.apply(input);
        }
        CompletableFuture<String> errorFuture = new CompletableFuture<>();
        errorFuture.completeExceptionally(new RuntimeException("Tool " + name + " not found in MCP registry."));
        return errorFuture;
    }
}
